di/getJRC_GEOPP_DB_csv.py

```diff
@@ -0,0 +1,137 @@
+selected_fields_spatial = selected_fields.copy()
+selected_fields_spatial.remove("FID")
+
+
+url = 'http://cidportal.jrc.ec.europa.eu/ftp/jrc-opendata/JRCOD/RES-DATA/10001/LATEST/JRC-GEOPP-DB.csv'
+
+
+def add_extras_to_schema(package):
+    # Add a new field to the first resource
+    package.pkg.descriptor["resources"][0]["schema"]["fields"].append(
+        dict(name="fields", type="object")
+    )
+    # Must yield the modified datapackage
+    yield package.pkg
+    # And its resources
+    yield from package
+
+
+def addFID(package):
+    # Add a new field to the first resource
+    package.pkg.descriptor["resources"][0]["schema"]["fields"].append(
+        dict(name="FID", type="string")
+    )
+    # Must yield the modified datapackage
+    yield package.pkg
+    # And its resources
+    yield from package
+    
+def fillFID(row):
+    row["FID"] = "jrc_geopp_db-"+row["id_powerplant"]
+
+def add_extras_column(row):
+    extra_fields = {}
+    for key,value in row.items():
+        if key not in selected_fields:
+            extra_fields[key] = value
+    row["fields"] = json.dumps(extra_fields)
+
+
+def retrieve(*args):
+    return Flow(
+        # Load inputs
+        load(url, format='csv', ),
+        
+        # Process them (if necessary)
+        update_package(name="jrc_geopp_db_csv"),
+        update_resource('JRC-GEOPP-DB', name='jrc_geopp_db', path='jrc_geopp_db.csv'),
+        add_extras_to_schema,
+        add_extras_column,
+        addFID,
+        fillFID,
+        # printer(num_rows=10),
+        select_fields(selected_fields, resources=None),
+        # printer(),
+        # add_computed_field(target=dict(name='geometry', type='geopoint'), operation='format', with_='{longitude}, {latitude}'),
+        add_computed_field(target=dict(name='geometry', type='string'), operation='format', with_='POINT({longitude} {latitude})'),
+        delete_fields(['latitude', 'longitude']),
+        
+        # # Update metadata
+        set_primary_key('id_powerplant'),
+        update_package(sources = [{
+                                    "title": "Joint Research Centre Data Catalogue",
+                                    "path": "https://data.jrc.ec.europa.eu/dataset/jrc-10128-10001"
+                                }]),
+        # printer(),
+        # Resource 1
+        duplicate(source="jrc_geopp_db", target_name="jrc_geopp_db_spatial", target_path="jrc_geopp_db_spatial.csv"),
+        delete_fields(selected_fields_spatial,resources=1),
+        # printer(num_rows=10),
+
+        # Resource 0
+        unpivot(unpivoting_fields, extra_keys, extra_value, regex=True,  resources=0),
+        delete_fields(['geometry'],resources=0),
+        find_replace([{"name": "value", "patterns":[{"find":"NULL","replace":""}]}], resources=0),
+        # update_schema(None, missingValues=["NULL", ""], resources=0),
+        set_type('value',type="number", resources=0),
+        # validate()
+        # printer(num_rows=10),
+    )
```

Comment:
that's a super long method definition that got me lost around line 66, any way that we could simplify this ?

Reply:
I will rewrite it using (Geo)Pandas only, should simplify things